File: 1206.design-skiplist.cpp
# 
# @lc app=leetcode id=1206 lang=cpp
# 
# [1206] Design Skiplist
# 
# @lc code=start
#include <vector>
#include <cstdlib>
#include <ctime>
using namespace std;

class Node {
public:
    int value;
    vector<Node*> forward;
    Node(int val, int level) : value(val), forward(level + 1, nullptr) {}
};

class Skiplist {
private:
    Node* head;
    int maxLevel;
    float probability;
    int randomLevel() {
        int lvl = 0;
        while ((rand() / double(RAND_MAX)) < probability && lvl < maxLevel) {
            lvl++;
        }
        return lvl;
    }
public:
    Skiplist() : maxLevel(16), probability(0.5) {
        srand(time(0));
        head = new Node(-1, maxLevel);
    }
    
bool search(int target) {
        Node* current = head;
        for (int i = maxLevel; i >= 0; i--) {
            while (current->forward[i] && current->forward[i]->value < target) {
                current = current->forward[i];
            }
        }
        current = current->forward[0];
        return current != nullptr && current->value == target;
    }
    
bool add(int num) {
g vector<Node*> update(maxLevel + 1);
g vector<Node*> update(maxLevel + 1);
d add(int num);current = head;current = head;for (int i = maxLevel; i >= 0; i--) {while (current->forward[i] && current->forward[i]->value < num) {current = current->forward[i];}update[i] = current;}int lvl = randomLevel();Node* newNode = new Node(num, lvl);for (int i = 0; i <= lvl; i++) {newNode->forward[i] = update[i]->forward[i];update[i]->forward[i] = newNode;}return true;
bool erase(int num) {
erase(num);bool erase(num);erase(num);bool erase(num);erase(num);erase(num);
erase(num);}

File: 1206.design-skiplist.go
#
# @lc app=leetcode id=1206 lang=golang
#
# [1206] Design Skiplist
#

# @lc code=start
import (
    "math/rand"
    "time"
)

const (
    maxLevel = 16
    p = 0.25
)

type Node struct {
    val int
    forward []*Node
}

type Skiplist struct {
    head *Node
    level int
}

func Constructor() Skiplist {
    rand.Seed(time.Now().UnixNano())
    head := &Node{val: -1, forward: make([]*Node, maxLevel)}
    return Skiplist{head: head, level: 1}
}

func randomLevel() int {
    lvl := 1
    for lvl < maxLevel && rand.Float64() < p {
        lvl++
    }
    return lvl
}

func (this *Skiplist) Search(target int) bool {
    curr := this.head
    for i := this.level - 1; i >= 0; i-- {
        for curr.forward[i] != nil && curr.forward[i].val < target {
            curr = curr.forward[i]
        }
    }
    curr = curr.forward[0]
    return curr != nil && curr.val == target
}

func (this *Skiplist) Add(num int) {
    update := make([]*Node, maxLevel)
    curr := this.head
    for i := this.level - 1; i >= 0; i-- {
        for curr.forward[i] != nil && curr.forward[i].val < num {
            curr = curr.forward[i]
        }
        update[i] = curr
    }
    lvl := randomLevel()
    if lvl > this.level {
        for i := this.level; i < lvl; i++ {
            update[i] = this.head
        }
        this.level = lvl
    }
    node := &Node{val: num, forward: make([]*Node, lvl)}
    for i := 0; i < lvl; i++ {
        node.forward[i] = update[i].forward[i]
        update[i].forward[i] = node
    }
}

func (this *Skiplist) Erase(num int) bool {
    update := make([]*Node, maxLevel)
    curr := this.head
    found := false
    for i := this.level - 1; i >= 0; i-- {
        for curr.forward[i] != nil && curr.forward[i].val < num {
            curr = curr.forward[i]
        }
        update[i] = curr
    }
    curr = curr.forward[0]
    if curr != nil && curr.val == num {
        found = true
        for i := 0; i < this.level; i++ {
            if update[i].forward[i] != curr {
                break
            }
            update[i].forward[i] = curr.forward[i]
        }
        for this.level > 1 && this.head.forward[this.level-1] == nil {
            this.level--
        }
    }
    return found
}

# @lc code=end